### ISSUE TITLE
[ci] Use NX caching to speed up CI

### ISSUE THREAD

**Issue by nolanlawson** [salesforce/lwc#3841]
We could be using the NX cache in CI (Nucleus, GitHub Actions) so speed up tests.

In the past, though, we've run into issues with the NX cache becoming out-of-date and causing hard-to-debug errors. So we should be careful to do this correctly.

This would speed up our builds by ~60 seconds per job.

**Comment by AllanOricil** [comment]
![image](https://github.com/salesforce/lwc/assets/55927613/2b5db7c1-f315-48ca-a736-70aa3c42e096)

:(

**Comment by AllanOricil** [comment]
@nolanlawson  
since caching isn't an option what if you guys use a more powerful machine for builds?

**Comment by AllanOricil** [comment]
I opened a feature request in the nx repo to see if they could create their own action to enable saving/restoring their cache during builds.

**Comment by AllanOricil** [comment]
https://github.com/nrwl/nx/issues/20327

**Comment by AllanOricil** [comment]
If there wasn't that issue with not being able to reuse cache hits from different machines, it could be possible to cache each task result as its own cache with a fixed key, and then update it every time nx detects a cache miss for that particular task

The workaround to update a github actions cache with a fixed key is here (I did not test, but there are people saying it works):
https://github.com/actions/cache/issues/342


we could read `./.nx-cache/run.json` to create a list of cache misses. Then we update the cache for each task that had a cache miss during a build using the above workaround.

**Comment by AllanOricil** [comment]
@nolanlawson 

Maybe you could get a free account?

![SmartSelect_20231118_021037_Brave.jpg](https://github.com/salesforce/lwc/assets/55927613/d822c5cc-e056-43d7-a6bf-75a0a2d4b9b8)

**Comment by AllanOricil** [comment]
There is an env variable to disable the unknow local cache 


NX_REJECT_UNKNOWN_LOCAL_CACHE=0

**Comment by AllanOricil** [comment]
> There is an env variable to disable the unknow local cache 
> 
> 
> NX_REJECT_UNKNOWN_LOCAL_CACHE=0


I think it would be possible to create a cache for each task run using the following key:

- sha256 of tarbal of the package folder (compute a hash from a tarbal with files sorted)
- yarn.lock

**Comment by nolanlawson** [comment]
I would be a bit uncomfortable turning this flag on. It looks like it's not recommended [for security reasons](https://nx.dev/recipes/troubleshooting/unknown-local-cache#you-share-cache-with-another-machine-using-a-network-drive).

The savings would also be <60s of build time per GitHub Action. I'm not sure it's worth the risk…

**Comment by AllanOricil** [comment]
> I would be a bit uncomfortable turning this flag on. It looks like it's not recommended [for security reasons](https://nx.dev/recipes/troubleshooting/unknown-local-cache#you-share-cache-with-another-machine-using-a-network-drive).
> 
> The savings would also be <60s of build time per GitHub Action. I'm not sure it's worth the risk…

I know it is just 60s, but I would like to point out that the security risks they mention wouldn't happen in Github Actions because each job run has its own working directory. It is not possible for one Job to overwrite the cache of the other one.
I red their concerns again and they tell people to avoid using the same "network drive" because one CI run could overwrite the .nx-cache of the other one. But as I said above, this won't happen in github actions. 

"Storing Nx's local cache on a network drive can present security risks. When a network drive is shared, every CI run has access to all the previously created Nx cache artifacts"

**Comment by AllanOricil** [comment]
I will try to implement an action specifically for nx caching. If I manage to put it together you could try it comparing the workflow results when using the action and when not. If they are the outputed builds are the same in bothe workflows we will know it worked. To verify they are the same we can compute their tarbal hashes.

**Comment by nolanlawson** [comment]
> I know it is just 60s, but I would like to point out that the security risks they mention wouldn't happen in Github Actions

Right... I guess what they are describing is the risk that the artifacts will be exposed on the network drive somehow. This is an OSS project, and the only secrets are the SauceLabs credentials, which shouldn't be written to the artifact cache by NX because there are no associated artifacts; we don't even run the karma/webdriver tests using `nx`.

So OK, maybe it is possible to turn on `NX_REJECT_UNKNOWN_LOCAL_CACHE=0` and not have any bad side effects. It looks like [we would not even be the only OSS project](https://github.com/search?q=NX_REJECT_UNKNOWN_LOCAL_CACHE+path%3A%2F%5E.github%5C%2Fworkflows%5C%2F%2F&type=code) doing this. I would just hope that it wouldn't break somehow in the future.
